python/lsst/daf/butler/formatters/parquet.py

```diff
@@ -21,177 +21,842 @@
+
+            filters = self.fileDescriptor.parameters.pop("filters", None)
+
+            if len(self.fileDescriptor.parameters):
+                raise ValueError(
+                    f"Unsupported parameters {self.fileDescriptor.parameters} in ArrowTable read."
+                )
+
+        metadata = schema.metadata if schema.metadata is not None else {}
+        arrow_table = pq.read_table(
+            self.fileDescriptor.location.path,
+            columns=par_columns,
+            use_threads=False,
+            use_pandas_metadata=(b"pandas" in metadata),
+            filters=filters,
+        )
+
+        return arrow_table
+
+    def write(self, inMemoryDataset: Any) -> None:
+        import numpy as np
+        import pandas as pd
+        from astropy.table import Table as astropyTable
+
+        location = self.makeUpdatedLocation(self.fileDescriptor.location)
+
+        if isinstance(inMemoryDataset, pd.DataFrame):
+            arrow_table = pandas_to_arrow(inMemoryDataset)
+        elif isinstance(inMemoryDataset, astropyTable):
+            arrow_table = astropy_to_arrow(inMemoryDataset)
+        elif isinstance(inMemoryDataset, np.ndarray):
+            arrow_table = numpy_to_arrow(inMemoryDataset)
+        elif isinstance(inMemoryDataset, pa.Table):
+            arrow_table = inMemoryDataset
         else:
-            raw_columns = list(self._splitColumnnNames(len(indexes), self.file.metadata.schema.names))
-            self.columns = pd.MultiIndex.from_tuples(raw_columns, names=[f["name"] for f in indexes])
-        self.indexLevelNames = tuple(self.columns.names)
+            raise ValueError("Unsupported type of inMemoryDataset for ParquetFormatter.")
+
+        pq.write_table(arrow_table, location.path)
+
+
+def arrow_to_pandas(arrow_table: pa.Table) -> pd.DataFrame:
+    """Convert a pyarrow table to a pandas DataFrame.
+    If the input table has ``pandas`` metadata in the schema it will be
+    used in the construction of the DataFrame.
+
+    Parameters
+    ----------
+    arrow_table : `pyarrow.Table`
+
+    Returns
+    -------
+    dataframe : `pandas.DataFrame`
+    """
+    return arrow_table.to_pandas(use_threads=False)
+
+
+def arrow_to_astropy(arrow_table: pa.Table) -> atable.Table:
+    """Convert a pyarrow table to an astropy.Table.
+
+    Parameters
+    ----------
+    arrow_table : `pyarrow.Table`
+
+    Returns
+    -------
+    table : `astropy.Table`
+    """
+    from astropy.table import Table
+
+    astropy_table = Table(arrow_to_numpy_dict(arrow_table))
+
+    metadata = arrow_table.schema.metadata if arrow_table.schema.metadata is not None else {}
+
+    _apply_astropy_metadata(astropy_table, metadata)
+
+    return astropy_table
+
+
+def arrow_to_numpy(arrow_table: pa.Table) -> np.ndarray:
+    """Convert a pyarrow table to a structured numpy array.
+
+    Parameters
+    ----------
+    arrow_table : `pyarrow.Table`
+
+    Returns
+    -------
+    array : `numpy.ndarray` (N,)
```

Comment:
Maybe add a short description of the shape of the numpy array here rather than using `N` where `N` has no context.